### PR TITLE
routing: inbound fees support for BuildRoute

### PR DESCRIPTION
**Pull request by bitromortac**
This PR attempts to add support for inbound fees to `BuildRoute`. The three passes `BuildRoute` goes through (see https://github.com/lightningnetwork/lnd/issues/8814#issuecomment-2160012218 for an explanation) are altered to determine the edges in the backward pass (from receiver to sender) as opposed to the previous approach where this is done in the forward pass (sender to receiver). This way we get fixed edges that can be used to determine the minimal sender/maximal receiver amount with inbound fees, doing a reverse amount calculation (thanks to @feelancer21 for the analysis https://github.com/lightningnetwork/lnd/issues/8814#issuecomment-2168351570). The computation for the outgoing amount based on the incoming amount is done using `big.Int` arithmetic.

Previous work on this: https://github.com/lightningnetwork/lnd/pull/7060

### CODE REVIEW

File: routing/router.go [hunk 1]
```diff
@@ -3254,50 +3259,102 @@ func backwardPass(unifiers []*edgeUnifier, useMinAmt bool,
...
+			)
+			if edge == nil {
+				log.Errorf("Cannot find policy with amt=%v "+
+					"for hop %v", runningAmt, i)
 
-			return nil, 0, ErrNoChannel{
-				position: i,
+				return nil, 0, ErrNoChannel{
+					position: i,
+				}
```
Comment (ProofOfKeags) [inline routing/router.go]: Split this case out of the loop, slice the edges down and run a single-branch loop for the remaining ones. We don't actually need to partition stuff since in both the outbound fee and inbound fee case there is only one exception and that exception is at one of the boundaries. This block will only ever be executed once, and it will be executed at the beginning.

File: routing/router.go [hunk 2]
```diff
@@ -3254,50 +3259,102 @@ func backwardPass(unifiers []*edgeUnifier, useMinAmt bool,
...
+			if edge == nil {
+				return nil, 0, ErrNoChannel{
+					position: i,
+				}
+			}
+
+			log.Tracef("Select channel %v at position %v",
+				edge.policy.ChannelID, i)
+
+			fee = int64(outboundFee)
```
Comment (ProofOfKeags) [inline routing/router.go]: It looks like you are converting to int64 only to convert immediately back to MSats.

File: routing/router.go [hunk 3]
```diff
@@ -3240,12 +3240,17 @@ func backwardPass(unifiers []*edgeUnifier, useMinAmt bool,
...
+	var nextIncoming lnwire.MilliSatoshi
 
 	// Traverse hops backwards to accumulate fees in the running amounts.
 	for i := len(unifiers) - 1; i >= 0; i-- {
-		localChan := i == 0
 		edgeUnifier := unifiers[i]
 
+		var (
+			edge *unifiedEdge
+			fee  int64
```
Comment (ProofOfKeags) [inline routing/router.go:3249]: I think it will make more sense to make this an MSat value.

File: routing/router.go [hunk 4]
```diff
@@ -3254,50 +3259,102 @@ func backwardPass(unifiers []*edgeUnifier, useMinAmt bool,
...
 	// been increased in the backward pass, fees need to be recalculated and
 	// amount ranges re-checked.
-	for i, edge := range unifiedEdges {
-		if i > 0 {
-			// Decrease the amount to send while going forward.
-			senderAmt -= edge.policy.ComputeFeeFromIncoming(
-				senderAmt,
-			)
-		}
+	for i := 1; i < len(unifiedEdges); i++ {
```
Comment (ProofOfKeags) [inline routing/router.go:3350]: 🫡

**Review by ProofOfKeags — CHANGES_REQUESTED** [review]
So this is my first time looking at this corner of the codebase, so the quality of my feedback may not be all that great but here is what I've come to understand.

The algorithm seems to be intrinsically imperative and essentially must be computed iteratively. Normally I hate this but in order to satisfy MinHTLC requirements I don't think there is a way around that. Given that the algorithm must be imperative, I think it is extremely important to be specific about the role of each variable. Unfortunately I'm not completely clear on precisely how we are computing things. My understanding is that there are broadly two cases that we need to consider:
1. We want to solve for the smallest amount we need to send to allow the receiver to successfully receive 1 sat.
2. We want to compute the amount that we need to send in order for the receiver to receive some other amount *ignoring MinHTLC requirements*. NOTE: I find it odd that we ignore MinHTLC requirements in this second case but it is a consequence of how the code is currently written, it's possible this wasn't intended and if so should be addressed.

From here It comes to my attention that the following expressions should hold:
```golang
sendAmt := recvAmt + fees
fees := outboundFees + inboundFees
outboundFees := fn.Sum(fn.Map(getOutboundFee, edges[1:]))
inboundFees := fn.Sum(fn.Map(getInboundFee, edges[:len(edges)-1]))
```

Separately, I have a few notes about naming things. Functions should be named according to the properties they satisfy, not based off of the procedure they follow. This is because the caller does not care about how a computation is carried out, they care about what computation is being done. Similarly, for variables, I'd *recommend* (though I do not *insist* on) naming them according to the eventual result they will produce in the case of accumulators, and for purely transient variables, you can drop things like "next" or "temp" and just favor using your character budget on being specific about what value is under that iterator.

I understand that a substantial portion of what I'm commenting on precedes this PR, however it is hard for me to comment on the quality of the diff without implicitly commenting on the quality of the existing codebase. Improving some of these things may be beyond the desired scope of this PR and in that case it's OK. However, in other parts of LND I've been favoring the approach of "no PR should make the quality of the codebase deteriorate".

Lmk if you think this is reasonable and attainable. If not, maybe we can converge on something that improves the codebase incrementally while not distracting us too much from the main feature goal.

File: routing/router.go [hunk 5]
```diff
@@ -3081,13 +3081,12 @@ func (r *ChannelRouter) MarkEdgeLive(chanID lnwire.ShortChannelID) error {
 // channels that satisfy all requirements.
 type ErrNoChannel struct {
 	position int
-	fromNode route.Vertex
```
[the remark this answers is not in the document]
Comment (guggero) [reply routing/router.go:3084]: It's mentioned in the commit message. So it sounds like we could extract that info at a later point, so we can log it? But in this commit it doesn't look like it yet. Still going through the PR, so will circle back to it.

File: routing/router.go [hunk 6]
```diff
@@ -3275,38 +3272,37 @@ func backwardPass(unifiers []*edgeUnifier, useMinAmt bool,
...
-// getPathEdges returns the edges that make up the path and the amount that gets
-// delivered to the receiver.
-func getPathEdges(receiverAmt lnwire.MilliSatoshi,
-	unifiers []*edgeUnifier,
-	bandwidthHints *bandwidthManager) ([]*unifiedEdge, lnwire.MilliSatoshi,
-	error) {
+// forwardPass returns the amount that a receiver will receive after deducting
+// all fees from the sender amount.
+func forwardPass(senderAmt lnwire.MilliSatoshi,
+	unifiedEdges []*unifiedEdge) (lnwire.MilliSatoshi, error) {
```
[the remark this answers is not in the document]
Comment (guggero) [reply routing/router.go:3332]: I don't think the naming is that bad here and for `backwardPass`. I think it's kind of important to the context whether you're going through the route forward or backward. So maybe just some additional info, perhaps `receiverAmountForwardPass` and `senderAmountBackwardPass`? Although that just makes things long and clunky...

File: routing/router.go (hunk 3 above)
Comment (guggero) [reply routing/router.go:3249]: Agree!

File: routing/router.go (hunk 4 above)
Comment (guggero) [reply routing/router.go:3350]: Why not just:
```
	for i := range unifiedEdges {
		outEdge := unifiedEdges[i]
		...
```

If this is about the loop variable bug thing?

File: docs/release-notes/release-notes-0.18.2.md [hunk 7]
```diff
@@ -59,6 +59,9 @@
 * [`ChanInfoRequest`](https://github.com/lightningnetwork/lnd/pull/8813)
   adds support for channel points.
 
+* [BuildRoute](https://github.com/lightningnetwork/lnd/pull/8886) now supports
```
Comment (guggero) [inline docs/release-notes/release-notes-0.18.2.md:62]: Need to move those to the 0.18.3 file now.

File: routing/router.go [hunk 8]
```diff
@@ -3363,3 +3363,49 @@ func forwardPass(senderAmt lnwire.MilliSatoshi,
...
 }
+
+// outgoingFromIncoming computes the outgoing amount based on the incoming
+// amount by subtracting fees from the incoming amount. Note that this is not
+// exactly the inverse of incomingFromOutgoing, because of the rounding.
+func outgoingFromIncoming(incomingAmt lnwire.MilliSatoshi,
+	incoming, outgoing *unifiedEdge) lnwire.MilliSatoshi {
+
+	// TODO: calculations with integer ceil/floor arithmetic
+	feeRateParts := 1e6
```
Comment (guggero) [inline routing/router.go:3375]: I think we can just keep everything multiplied by a million and then do integer division later on?
Not sure about the requirements of the actual algorithm, I just know we should try to avoid `float64` wherever possible!
But I guess that's exactly what the TODOs here are for.

**Review by guggero — COMMENTED** [review]
Did a first pass to load context. Refactor work looks good to me :+1: 
But looks like the actual calculation has some TODOs left, so will revisit once ready.

File: routing/router.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply routing/router.go:3332]: My point is that the name of a function serves as a way to inform the caller about what the thing does. We don't actually need it to tell us how it does it. By the time we are looking at the function's internals we have plenty of other mechanisms to explain to ourselves how it is accomplishing that. The caller only has:
1. the name
2. the godoc
3. the type signature

The detail of which direction we iterate in order to make the determination is of no importance to the user of this function. The user calls it to determine the receiverAmount. Could also name it `deductFees` or something to that effect.

File: routing/router.go [hunk 9]
```diff
@@ -3171,11 +3178,9 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
 }
 
 // getRouteUnifiers returns a list of edge unifiers for the given route.
-func getRouteUnifiers(source route.Vertex, hops []route.Vertex,
-	useMinAmt bool, runningAmt lnwire.MilliSatoshi,
-	outgoingChans map[uint64]struct{}, graph routingGraph,
-	bandwidthHints *bandwidthManager) ([]*edgeUnifier, lnwire.MilliSatoshi,
-	error) {
+func getEdgeUnifiers(source route.Vertex, hops []route.Vertex,
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go:1501]: The graph in LN pathfinding (at least in LND) is regarded as a simple graph, but the LN graph can have multiple channels between nodes, with different (and maybe conflicting) channel policies and constraints (min/max HTLC sizes). This is why we unify those channels and take worst case boundaries into account in order to improve on the probability that a payment can successfully be routed. For example you'd take the maximal fee that's advertised. Here we only collect the unifiers, comprising the set of policies for each node pair, such that we can dynamically decide which policy to use for a specific amount.

File: routing/router.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go:1396]: I removed it because the node pubkey isn't available later when encountering an error and it's redundant, one can figure out the problematic pub key by the index of the input keys. I could pass in the hops into the functions that generate those errors, to remain backward compatible if wanted?

File: routing/router.go [hunk 10]
```diff
@@ -3243,19 +3261,17 @@ func getRouteUnifiers(source route.Vertex, hops []route.Vertex,
...
-// including fees, to send the payment.
-func getPathEdges(source route.Vertex, receiverAmt lnwire.MilliSatoshi,
-	unifiers []*edgeUnifier, bandwidthHints *bandwidthManager,
-	hops []route.Vertex) ([]*unifiedEdge,
-	lnwire.MilliSatoshi, error) {
+// getPathEdges returns the edges that make up the path and the amount that gets
+// delivered to the receiver.
+func getPathEdges(receiverAmt lnwire.MilliSatoshi,
+	unifiers []*edgeUnifier,
+	bandwidthHints *bandwidthManager) ([]*unifiedEdge, lnwire.MilliSatoshi,
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go]: Right I agree, but if I'd change the function parameter type to `bandwidthManager`, it wouldn't fulfill the `bandwidthHints` interface in `getEdge` anymore. Looking at bandwidth hints, it looks like implementing it without a pointer receiver would work, but not really sure I should do that in this PR?

File: routing/router.go [hunk 11]
```diff
@@ -3130,31 +3130,45 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
...
-	pathEdges, receiverAmt, err := getPathEdges(
-		senderAmt, unifiers, bandwidthHints,
-	)
-	if err != nil {
-		return nil, err
+		pathEdges, receiverAmt, err = getPathEdges(
+			senderAmt, unifiers, bandwidthHints,
+		)
+		if err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go]: Yes was a bit strange to me to add it :), the intention here was to have a cleaner diff in the next commit.

File: routing/router.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go]: I agree that things should be named by what they do, not how :+1:. I was just led to the names by their symmetry :). On the other hand, as Oli wrote, also for me it's kind of more intuitive to have that association with the direction. I took his suggestion to use mixed names

File: routing/router.go (hunk 3 above)
Comment (bitromortac) [reply routing/router.go]: Done!

File: routing/router.go (hunk 1 above)
Comment (bitromortac) [reply routing/router.go]: I chose it that way because `newRoute` has the same structure to have an easy comparison with it as the reference implementation. I'm aware that there is some code repetition, Joost tried to solve this as well in previous draft PRs, but generally I think that the current approach decouples the concerns of finding unified edges, which is not included in `newRoute`. The minimal amount search is a bit akin to do a first pathfinding stage that ends up with calling `newRoute` (this is how it's done in `findPath` as well) and kind of preexisting. Happy to change this to reduce the loop complexity if wanted

File: routing/router.go (hunk 4 above)
Comment (bitromortac) [reply routing/router.go:1665]: > If this is about the loop variable bug thing?

this loop starts with index one, not sure you saw that?

File: routing/router.go (hunk 8 above)
Comment (bitromortac) [reply routing/router.go]: It was kind of tough to do integer arithmetic here because formulas involve products of PPM*PPM*amount, which can easily overflow a int64 and we need to handle negative numbers as well. I tried an approach with floats, why at least don't have a problem overflowing when adding/multiplying numbers and I think there are no problems related to accumulating errors, but not too knowledgeable here. Another option would be to use `big.Int`, but the syntax there looks very clumsy

**Comment by bitromortac** [comment]
Thanks for the reviews @guggero and @ProofOfKeags and also for the improvement suggestions! 

> My understanding is that there are broadly two cases that we need to consider

Yes pretty well described :+1: 

> The algorithm seems to be intrinsically imperative and essentially must be computed iteratively.
> From here It comes to my attention that the following expressions should hold
```
sendAmt := recvAmt + fees
fees := outboundFees + inboundFees
outboundFees := fn.Sum(fn.Map(getOutboundFee, edges[1:]))
inboundFees := fn.Sum(fn.Map(getInboundFee, edges[:len(edges)-1]))
```

Roughly yes, unfortunately the inbound fees are coupled to the previous hop's outbound fees, which is why I think the map approach wouldn't work well.

I added some documentation to the amount calculation, hope that helps. I'm still using float numerics, but need to look into that again to be sure it is precise and what the boundaries of failures would be.

> However, in other parts of LND I've been favoring the approach of "no PR should make the quality of the codebase deteriorate".

Totally agree, happy to address things!

File: routing/router.go [hunk 12]
```diff
@@ -1535,67 +1570,228 @@ func getRouteUnifiers(source route.Vertex, hops []route.Vertex,
...
+		return incomingAmt
+	}
+
+	// This is the second case where we have to compute the outgoing amount.
+	if lnwire.MilliSatoshi(testAmtF) > incomingAmt {
+		// Compute the exact outgoing amount as a float.
+		out := (A - n) / m
+
+		// We round to PPM precision.
+		rounded := math.Round(out*PPM) / PPM
```
Comment (bitromortac) [inline routing/router.go:1784]: will need to check if this is needed

File: routing/router.go [hunk 13]
```diff
@@ -1620,7 +1618,13 @@ func senderAmtBackwardPass(unifiers []*edgeUnifier, useMinAmt bool,
...
 			}
 
-			fee = outboundFee
+			inboundFee := calcCappedInboundFee(
+				edge, netAmount, outboundFee,
+			)
+
+			fee = lnwire.MilliSatoshi(
+				int64(outboundFee) + inboundFee,
+			)
```
Comment (ProofOfKeags) [inline routing/router.go:1627]: 🫡

**Review by ProofOfKeags — CHANGES_REQUESTED** [review]
Much improved. I still have some stuff I'd like to see addressed, particularly around testing and floating point arithmetic.

File: routing/router.go [hunk 14]
```diff
@@ -1675,3 +1676,119 @@ func receiverAmtForwardPass(runningAmt lnwire.MilliSatoshi,
...
+	// Convert all quantities to floats to be able to hande negative values.
+	// The formulas to compute the outgoing amount involve terms with
+	// PPM*PPM*A, which can easily overflow an int64. Other terms may
+	// overflow as well, which is why we chose to work with float64 here.
+	A := float64(incomingAmt)
+	Ro := float64(outgoing.policy.FeeProportionalMillionths)
+	Bo := float64(outgoing.policy.FeeBaseMSat)
+	Ri := float64(incoming.inboundFees.Rate)
+	Bi := float64(incoming.inboundFees.Base)
+	PPM := float64(1_000_000)
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go:1698]: I tried using int64, but there are terms in the formulas like PPM*PPM*amt_msat, where PPM*PPM is already on the order of `math.MaxInt64`. I'll try to change this using `big.Int`, but it won't be as readable I think.

File: routing/router_test.go [hunk 15]
```diff
@@ -1852,6 +1852,175 @@ func TestSenderAmtBackwardPass(t *testing.T) {
...
+	if int64(outgoingFee)+inboundFee < 0 {
+		return outgoingAmt
+	}
+
+	return netAmount + lnwire.MilliSatoshi(inboundFee)
+}
+
+// TestInboundOutbound tests the functions that computes the incoming and
+// outgoing amounts based on the fees of the incoming and outgoing channels.
+func TestInboundOutbound(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router_test.go:1951]: Right, I had property-based testing in mind when outsourcing to `testInboundOutboundFee`. I added `FuzzInboundOutboundFee` and have thought about using the quick library, but didn't use it so far as a predicate function would not use `require` to indicate which property failed exactly. I'll do the switch to `quick.Check`, would be interested in the differences between fuzzing and `quick.Check`. Fuzzing isn't run in the standard unit tests, how many parameters does `quick.Check` run? Also, does it only check random inputs or does it do some code analysis to create better edge cases, need to read up on that :)

File: routing/router.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go]: Ah, sorry, was confused, changed it to the interface type, originally I tried removing the pointer :+1:

File: routing/router_test.go [hunk 16]
```diff
@@ -1852,6 +1852,175 @@ func TestSenderAmtBackwardPass(t *testing.T) {
...
+	inboundFee := incoming.inboundFees.CalcFee(netAmount)
+
+	// The inbound fee is not allowed to reduce the incoming amount below
+	// the outgoing amount.
+	if int64(outgoingFee)+inboundFee < 0 {
+		return outgoingAmt
+	}
+
+	return netAmount + lnwire.MilliSatoshi(inboundFee)
+}
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router_test.go:1947]: Initially I had it next to it, but it's only used for testing, so not sure what the best location is

File: routing/router.go [hunk 17]
```diff
@@ -1675,3 +1676,119 @@ func receiverAmtForwardPass(runningAmt lnwire.MilliSatoshi,
...
+	// total node fee (Ai - Ao) is non-negative. This is commonly enforced
+	// by routing nodes.
+
+	// The function Ai(Ao) is given by:
+	// Ai(Ao) = (Ao + Bo + Ro/PPM) + (Bi + (Ao + Ro/PPM + Bo)*Ri/PPM), where
+	// the first term is the net amount (the outgoing amount plus the
+	// outbound fee), and the second is the inbound fee computed based on
+	// the net amount.
+
+	// Ai(Ao) can potentially become more negative in amplitude than Ao,
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go:1715]: right, I probably had `magnitude` in mind :), changed to `absolute value`

File: routing/router.go [hunk 18]
```diff
@@ -1675,3 +1676,119 @@ func receiverAmtForwardPass(runningAmt lnwire.MilliSatoshi,
...
+	//
+	// We know Ai(A) = A * m + n, therefore Ai(A) > A.
+	//
+	// This means that if we apply the incoming amount calculation to the
+	// **incoming** amount, and this condition holds, then we know that we
+	// deal with the second case, being able to compute the outgoing amount
+	// based off the formula Ao = (A - n) / m, otherwise we will just return
+	// the incoming amount.
+
+	// In case the inbound fee rate is more negative than 100%, we fail to
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go:1759]: it is a fee rate, so it would mean a number below -1 or -100%, I corrected it to `less than -1 (-100%)`, thanks!

File: routing/router.go [hunk 19]
```diff
@@ -1675,3 +1676,119 @@ func receiverAmtForwardPass(runningAmt lnwire.MilliSatoshi,
...
+	// The function Ai(Ao) is given by:
+	// Ai(Ao) = (Ao + Bo + Ro/PPM) + (Bi + (Ao + Ro/PPM + Bo)*Ri/PPM), where
+	// the first term is the net amount (the outgoing amount plus the
+	// outbound fee), and the second is the inbound fee computed based on
+	// the net amount.
+
+	// Ai(Ao) can potentially become more negative in amplitude than Ao,
+	// which is why the above mentioned capping is needed. We can abbreviate
+	// Ai(Ao) with Ai(Ao) = m*Ao + n, where m and n are:
+	// m := (1 + Ro/PPM) * (1 + Ri/PPM)
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go:1718]: not sure I follow, could you explain that?

File: routing/router_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply routing/router_test.go:1976]: it's fine to have functions that are only used in testing if they are good API additions. This is how we build up library infrastructure and can more effectively react to changing requirements. Even if it is unused in LND itself, I'd put it there. You never know what's going to happen when someone comes into the code next, and then needs that function only to not notice that it's hidden in the test files.

File: routing/router.go [hunk 20]
```diff
@@ -1749,36 +1750,97 @@ func outgoingFromIncoming(incomingAmt lnwire.MilliSatoshi,
 	// In case the inbound fee rate is less than -1 (-100%), we fail to
 	// compute the outbound amount and return the incoming amount. This also
 	// protects against zero division later.
-	m := (PPM*PPM + Ri*PPM + Ro*PPM + Ro*Ri) / (PPM * PPM)
-	if m <= 0 {
+	// m := (PPM*PPM + Ri*PPM + Ro*PPM + Ro*Ri) / (PPM * PPM)
+
+	// Note: this could maybe fit into an int64.
+	m1 := new(big.Int).Mul(PPM, PPM)
```
Comment (ProofOfKeags) [inline routing/router.go:1750]: I think this reads fine. However if you *want* to improve it you can always postfix these lines with their more familiar expressions like this:

```golang
m1 := new(big.Int).Mul(PPM, PPM) // m1 := PPM * PPM
...
// m := m1 + m2 + m3 + m4
m := big.NewInt(0)
m.Add(m, m1)
...
m.Add(m, m4)
```

File: routing/router.go [hunk 21]
```diff
@@ -1749,36 +1750,97 @@ func outgoingFromIncoming(incomingAmt lnwire.MilliSatoshi,
...
-		numerator := PPM * (PPM*(A-Bo-Bi) - Bo*Ri)
-		denominator := PPM*(PPM+Ri+Ro) + Ri*Ro
-		out := numerator / denominator
-
-		// We round to PPM precision.
-		rounded := math.Round(out*PPM) / PPM
+	if lnwire.MilliSatoshi(testAmt.Int64()) > incomingAmt {
+		// Compute the outgoing amount in terms of big.Int.
+		// numerator := PPM * (PPM*(A-Bo-Bi) - Bo*Ri)
+		// denominator := PPM*(PPM+Ri+Ro) + Ri*Ro
```
Comment (ProofOfKeags) [inline routing/router.go:1806]: I know that golang generally doesn't do this for whatever reason, but I would always put spaces between the operators and their operands.

File: routing/router.go [hunk 22]
```diff
@@ -1439,45 +1439,33 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
...
-			return nil, err
-		}
-
-		receiverAmt = *amt
 	}
 
+	// Otherwise, we just use the requested amount as the receiver amount.
+	amt.WhenSome(func(amt lnwire.MilliSatoshi) {
+		receiverAmt = amt
+	})
```
Comment (ProofOfKeags) [inline routing/router.go]: You can do this, but it'd be better if you did this instead:

```golang
receiverAmt, err := fn.ElimOption(
        amt,
        func() {
                receiverAmt, err := receiverAmtForwardPass(senderAmt, pathEdges)
                return NewResult(receiverAmt, err)
        }
        Ok[lnwire.MilliSatoshi],
).Unpack()
```

File: routing/router.go [hunk 23]
```diff
@@ -1543,9 +1531,9 @@ func getEdgeUnifiers(source route.Vertex, hops []route.Vertex,
 
 // senderAmtBackwardPass returns a list of unified edges for the given route and
 // determines the amount that should be sent to fulfill min HTLC requirements.
-// The minimal sender amount can be searched for by activating useMinAmt.
-func senderAmtBackwardPass(unifiers []*edgeUnifier, useMinAmt bool,
-	receiverAmt lnwire.MilliSatoshi,
+// The minimal sender amount can be searched for by using amt=None.
+func senderAmtBackwardPass(unifiers []*edgeUnifier,
+	amt fn.Option[lnwire.MilliSatoshi],
```
Comment (ProofOfKeags) [inline routing/router.go:1536]: 🫡

File: routing/router_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router_test.go]: Moved it back

File: routing/router.go [hunk 24]
```diff
@@ -1441,31 +1441,45 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
...
-	}
-
-	senderAmt, err := senderAmtBackwardPass(
-		unifiers, useMinAmt, runningAmt, bandwidthHints,
-	)
-	if err != nil {
-		return nil, err
-	}
+		senderAmt, err := senderAmtBackwardPass(
+			unifiers, useMinAmt, *amt, bandwidthHints,
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go]: I added a commit that changes to an `Option`, thanks for the suggestion and help with that!

File: routing/router.go (hunk 20 above)
Comment (bitromortac) [reply routing/router.go:1783]: I added some comments to make the calculations hopefully easier to follow

File: routing/router.go (hunk 21 above)
Comment (bitromortac) [reply routing/router.go]: right, I just commented the auto-formatted lines, imo it's best to read if multiplication/division doesn't have spaces, but addition/subtraction has, change it that way

File: routing/router.go (hunk 12 above)
Comment (bitromortac) [reply routing/router.go]: This is not needed anymore due to integer arithmetic.

File: routing/router.go (hunk 1 above)
Comment (bitromortac) [reply routing/router.go]: Thanks for spelling that out, done!

**Review by ProofOfKeags — COMMENTED** [review]
Super close at this point.

File: routing/router.go [hunk 25]
```diff
@@ -1545,69 +1545,121 @@ func getEdgeUnifiers(source route.Vertex, hops []route.Vertex,
...
 
-	// Traverse hops backwards to accumulate fees in the running amounts.
-	for i := len(unifiers) - 1; i >= 0; i-- {
-		localChan := i == 0
-		edgeUnifier := unifiers[i]
+	// We traverse the route backwards and handle the last hop separately.
+	edgeUnifier := unifiers[len(unifiers)-1]
+
+	// The last hop only forwards the amount that the destination should
+	// receive.
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go]: Yes, I think the documentation was not that great, added some more in the respective places.

File: routing/router_test.go [hunk 26]
```diff
@@ -1881,6 +1881,153 @@ func TestSenderAmtBackwardPass(t *testing.T) {
...
+	// We check that up to rounding the amounts are equal.
+	require.InDelta(
+		t, int64(outgoingAmt), int64(outgoingAmtNew), 1.0, debugStr,
+	)
+
+	// If we round, the computed outgoing amount should be larger than the
+	// exact outgoing amount, to not hit any min HTLC limits.
+	require.GreaterOrEqual(
+		t, int64(outgoingAmtNew), int64(outgoingAmt), debugStr,
+	)
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router_test.go]: I added some distinguishing strings next to the `debugStr`

File: routing/router.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go:1730]: Hm, I'm still a bit confused by the statement, since I'd rather say it's accurate in real space (up to a cast to compare to the int result) but not in integer space as you need to do an integer division twice with the accompanied rounding, but good it's resolved

File: routing/router.go [hunk 27]
```diff
@@ -1545,69 +1545,119 @@ func getEdgeUnifiers(source route.Vertex, hops []route.Vertex,
...
+				incomingAmt = min
 			}
 		}
 
-		// Get an edge for the specific amount that we want to forward.
-		edge := edgeUnifier.getEdge(runningAmt, bandwidthHints, 0)
-		if edge == nil {
-			log.Errorf("Cannot find policy with amt=%v for hop %v",
-				runningAmt, i)
+		// A --current hop-- B --next hop: incomingAmt-- C
```
Comment (ProofOfKeags) [inline routing/router.go:1594]: love this comment

File: routing/router.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (ProofOfKeags) [reply routing/router.go:1730]: I thought the reason we add the 1 in the first place is because of the truncation of the remainder in integer space. By keeping it in real space I do not believe we need the `1 +` part of the expression. Maybe I misunderstand.

File: routing/router.go [hunk 28]
```diff
@@ -1439,43 +1439,38 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
...
+	// just use the requested amount.
+	receiverAmt, err = fn.ElimOption(
+		amt,
+		func() fn.Result[lnwire.MilliSatoshi] {
+			receiverAmt, err := receiverAmtForwardPass(
+				senderAmt, pathEdges,
+			)
 
-		receiverAmt = *amt
+			return fn.NewResult(receiverAmt, err)
```
Comment (ProofOfKeags) [inline routing/router.go]: I learned from @Roasbeef that you can actually just do this and it mysteriously works:
```golang
func() fn.Result[lnwire.MilliSatoshi] {
        return fn.NewResult(
                receiverAmtForwardPass(
                        senderAmt, pathEdges,
                )
        )
}
```

**Review by ProofOfKeags — APPROVED** [review]

File: routing/router.go [hunk 29]
```diff
@@ -1481,12 +1488,10 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
...
 
-// getRouteUnifiers returns a list of edge unifiers for the given route.
-func getRouteUnifiers(source route.Vertex, hops []route.Vertex,
-	useMinAmt bool, runningAmt lnwire.MilliSatoshi,
-	outgoingChans map[uint64]struct{}, graph Graph,
-	bandwidthHints *bandwidthManager) ([]*edgeUnifier, lnwire.MilliSatoshi,
-	error) {
+// getEdgeUnifiers returns a list of edge unifiers for the given route.
+func getEdgeUnifiers(source route.Vertex, hops []route.Vertex,
+	outgoingChans map[uint64]struct{},
```
[the remark this answers is not in the document]
Comment (guggero) [reply routing/router.go:1498]: I agree, a set would be nice here. But changing that would also require a change in `newNodeEdgeUnifier` which is used in a couple of places. So non-blocking IMO, don't want to turn this into a big type refactor PR.

File: routing/router.go [hunk 30]
```diff
@@ -1545,69 +1545,119 @@ func getEdgeUnifiers(source route.Vertex, hops []route.Vertex,
...
 	error) {
 
 	var unifiedEdges = make([]*unifiedEdge, len(unifiers))
 
-	// Traverse hops backwards to accumulate fees in the running amounts.
-	for i := len(unifiers) - 1; i >= 0; i-- {
-		localChan := i == 0
-		edgeUnifier := unifiers[i]
+	// We traverse the route backwards and handle the last hop separately.
+	edgeUnifier := unifiers[len(unifiers)-1]
```
Comment (guggero) [inline routing/router.go:1550]: Do we need a bounds check here or are we certain there's always at least one edge?

File: routing/router_test.go [hunk 31]
```diff
@@ -1881,6 +1881,156 @@ func TestSenderAmtBackwardPass(t *testing.T) {
...
+			outgoingBase: 20,
+			outgoingRate: 100,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+
+		t.Run(tt.name, func(t *testing.T) {
+			testInboundOutboundFee(t,
```
Comment (guggero) [inline routing/router_test.go:2025]: micro-nit: `t` on new line. Also, usually we use `tc` for test cases and `tt` for sub `*testing.T` within a `t.Run()`. But really inconsequential, feel free to ignore.

**Review by guggero — COMMENTED** [review]
Looks very nice!
Two questions and a nit, other than that LGTM :tada:

File: routing/router.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (bitromortac) [reply routing/router.go:1730]: It's not about rounding here, it's the exact formula, but factored to see how the fee rates compare to 1, to see if it results in a zero or negative overall term, see the og derivation https://github.com/feelancer21/lnd/blob/f6f05fa930985aac0d27c3f6681aada1b599162a/prototype/doc.md#outgoing---incoming.

File: routing/router.go (hunk 30 above)
Comment (bitromortac) [reply routing/router.go:1552]: nice, I added a check here and on the RPC level (new commit)

File: routing/router_test.go (hunk 31 above)
Comment (bitromortac) [reply routing/router_test.go]: cool, I never realized that pattern, updated :pray:

**Review by guggero — APPROVED** [review]
LGTM :tada:

**Comment by bitromortac** [comment]
I noticed that when constructing the edge unifiers, we always included inbound fees in the policy. In pathfinding we don't set them to be active for the last hop, which is important for edge constraint checks (updated and documented in `TestSenderAmtBackwardPass`). I also added a test case for a min amount search with inbound fees that also displays an interesting edge case with rounding, which gets corrected in `newRoute`. I think that an itest could be interesting here as well, on the otherhand, because everything passes through `newRoute`, this should be covered by the pathfinding tests.
